### ISSUE TITLE
Trait impl conflict behavior differs for sized and unsized types

### ISSUE THREAD

**Issue by joshlf** [rust-lang/rust#104297]
I was writing a test macro to confirm that a type didn't implement a trait. For an invocation `assert_not_impl!(Type: Trait)`, it emits code like:

```rust
const _: () = {
    trait Blanket {}
    impl<T: Trait> Blanket for T {}
    impl Blanket for Type {}
};
```

The goal was to cause a trait impl conflict if the `Trait` was spuriously implemented. I found that while this worked for sized types, it didn't catch the error for unsized types. In other words, the following code fails to compile:

```rust
struct Foo;

trait Bar {}

impl Bar for Foo {}

const _: () = {
    trait Blanket {}
    impl<T: Bar> Blanket for T {}
    impl Blanket for Foo {}
};
```

However, if we modify `Foo` to be an unsized type, the code compiles successfully:

```rust
struct Foo([u8]);

trait Bar {}

impl Bar for Foo {}

const _: () = {
    trait Blanket {}
    impl<T: Bar> Blanket for T {}
    impl Blanket for Foo {}
};
```

I'm not convinced that this is a bug, but it's certainly surprising, and might point to a bug.

## Version
1.65.0

**Comment by compiler-errors** [comment]
Pretty sure that this is due to the implicit `T: Sized` bound that you get on `impl<T: Bar> Blanket for T {}`. You probably want to emit a blanket impl with a `?Sized` on the parameter to opt-out of this bound.

**Comment by joshlf** [comment]
You're right! False alarm. Thanks!
